File: packages/server/src/workspaces/manager.ts
import path from "path"
import os from "os"
import { createHash } from "crypto"
import { existsSync, mkdirSync } from "fs"
import { execSync, spawnSync } from "child_process"
import { connect } from "net"
import { EventBus } from "../events/bus"
import type { SettingsService } from "../settings/service"
import type { BinaryResolver } from "../settings/binaries"
import { FileSystemBrowser } from "../filesystem/browser"
import { searchWorkspaceFiles, WorkspaceFileSearchOptions } from "../filesystem/search"
import { clearWorkspaceSearchCache } from "../filesystem/search-cache"
import { WorkspaceDescriptor, WorkspaceFileResponse, FileSystemEntry } from "../api-types"
import { WorkspaceRuntime, ProcessExitInfo } from "./runtime"
import { Logger } from "../logger"
import { getOpencodeConfigDir } from "../opencode-config.js"
import { BIN_DIR, BINARY_NAME, triggerBinaryDownload, resolveBinaryPathFromUserShell } from "../opencode-paths"
import {
  OPENCODE_SERVER_PASSWORD_ENV,
  OPENCODE_SERVER_USERNAME_ENV,
  resolveOpencodeServerAuth,
} from "./opencode-auth"

const STARTUP_STABILITY_DELAY_MS = 1500



interface WorkspaceManagerOptions {
  rootDir: string
  settings: SettingsService
  binaryResolver: BinaryResolver
  eventBus: EventBus
  logger: Logger
  getServerBaseUrl: () => string
  /** Optional CA bundle path to trust EmbeddedCowork HTTPS certs. */
  nodeExtraCaCertsPath?: string
}

interface WorkspaceRecord extends WorkspaceDescriptor {}

export class WorkspaceManager {
  private readonly workspaces = new Map<string, WorkspaceRecord>()
  private readonly runtime: WorkspaceRuntime
  private readonly opencodeConfigDir: string
  private readonly opencodeAuth = new Map<string, { username: string; password: string; authorization: string }>()

  constructor(private readonly options: WorkspaceManagerOptions) {
    this.runtime = new WorkspaceRuntime(this.options.eventBus, this.options.logger)
    this.opencodeConfigDir = getOpencodeConfigDir()
  }

  list(): WorkspaceDescriptor[] {
    return Array.from(this.workspaces.values())
  }

  get(id: string): WorkspaceDescriptor | undefined {
    return this.workspaces.get(id)
  }

  getInstancePort(id: string): number | undefined {
    return this.workspaces.get(id)?.port
  }

  getInstanceAuthorizationHeader(id: string): string | undefined {
    return this.opencodeAuth.get(id)?.authorization
  }

  listFiles(workspaceId: string, relativePath = "."): FileSystemEntry[] {
    const workspace = this.requireWorkspace(workspaceId)
    const browser = new FileSystemBrowser({ rootDir: workspace.path })
    return browser.list(relativePath)
  }

  searchFiles(workspaceId: string, query: string, options?: WorkspaceFileSearchOptions): FileSystemEntry[] {
    const workspace = this.requireWorkspace(workspaceId)
    return searchWorkspaceFiles(workspace.path, query, options)
  }

  readFile(workspaceId: string, relativePath: string): WorkspaceFileResponse {
    const workspace = this.requireWorkspace(workspaceId)
    const browser = new FileSystemBrowser({ rootDir: workspace.path })
    const contents = browser.readFile(relativePath)
    return {
      workspaceId,
      relativePath,
      contents,
    }
  }

  writeFile(workspaceId: string, relativePath: string, contents: string): void {
    const workspace = this.requireWorkspace(workspaceId)
    const browser = new FileSystemBrowser({ rootDir: workspace.path })
    browser.writeFile(relativePath, contents)
  }

  /**
   * Compute a deterministic workspace ID based on the folder's content identity.
   *
   * - If the folder is inside a git repository: ID is derived from the git root
   *   commit hash + the folder's relative path from git root. This ensures the
   *   ID stays the same when the folder is moved/renamed within the repo.
   * - Otherwise: fall back to hashing the absolute path.
   */
  private computeWorkspaceId(folder: string): string {
    const absPath = path.resolve(folder)

    // Attempt git-based identity
    try {
      const topLevel = execSync("git rev-parse --show-toplevel", { cwd: absPath, encoding: "utf8", timeout: 5000 }).trim()
      const rootHash = execSync("git rev-list --max-parents=0 HEAD", { cwd: absPath, encoding: "utf8", timeout: 5000 })
        .trim()
        .split("\n")
        .filter(Boolean)
        .sort()[0]
      if (rootHash) {
        const relPath = path.relative(topLevel, absPath) || "/"
        return createHash("sha256").update(`${rootHash}:${relPath}`).digest("hex").slice(0, 12)
      }
    } catch {
      // Not a git repo or git unavailable — fall through
    }

    // Fallback: use absolute path hash
    return createHash("sha256").update(absPath).digest("hex").slice(0, 12)
  }

  async create(folder: string, name?: string): Promise<WorkspaceDescriptor> {
 
    const workspacePath = path.isAbsolute(folder) ? folder : path.resolve(this.options.rootDir, folder)
    const id = this.computeWorkspaceId(workspacePath)
    const binary = this.options.binaryResolver.resolveDefault()
    let resolvedBinaryPath = this.resolveBinaryPath(binary.path)
    clearWorkspaceSearchCache(workspacePath)

    this.options.logger.info({ workspaceId: id, folder: workspacePath, binary: resolvedBinaryPath }, "Creating workspace")

    const proxyPath = `/workspaces/${id}/worktrees/root/instance`


    const descriptor: WorkspaceRecord = {
      id,
      path: workspacePath,
      name,
      status: "starting",
      proxyPath,
      binaryId: resolvedBinaryPath,
      binaryLabel: binary.label,
      binaryVersion: binary.version,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    }

    this.workspaces.set(id, descriptor)


    this.options.eventBus.publish({ type: "workspace.created", workspace: descriptor })

    if (!existsSync(resolvedBinaryPath)) {
      this.options.logger.info({ workspaceId: id, binary: resolvedBinaryPath }, "Binary not found, waiting for download")
      await triggerBinaryDownload(this.options.logger)
      resolvedBinaryPath = this.resolveBinaryPath(binary.path)
      if (!existsSync(resolvedBinaryPath)) {
        throw new Error(`OpenCode binary still not found after auto-download: ${resolvedBinaryPath}`)
      }
      descriptor.binaryId = resolvedBinaryPath
      descriptor.binaryLabel = "auto-downloaded"
      this.options.logger.info({ workspaceId: id, path: resolvedBinaryPath }, "Binary ready after auto-download")
    }

    const serverConfig = this.options.settings.getOwner("config", "server")
    const envVars = (serverConfig as any)?.environmentVariables
    const userEnvironment = envVars && typeof envVars === "object" && !Array.isArray(envVars) ? (envVars as any) : {}

    const opencodeAuth = resolveOpencodeServerAuth({ userEnvironment })
    this.opencodeAuth.set(id, opencodeAuth)

    // Determine session database path based on user preference.
    //   - "project": place DB inside the workspace folder so it travels with the project
    //   - "global":  place DB in a centralized directory keyed by workspaceId
    const sessionStorageMode = (serverConfig as any)?.sessionStorageMode ?? "project"
    let dbPath: string
    if (sessionStorageMode === "global") {
      dbPath = path.join(os.homedir(), ".embeddedcowork", "session-data", `${id}.db`)
    } else {
      dbPath = path.join(workspacePath, ".embeddedcowork", "session", "data.db")
    }
    // Ensure parent directory exists before opencode starts; Database(path, { create: true })
    // only creates the file, not intermediate directories.
    mkdirSync(path.dirname(dbPath), { recursive: true })

    const environment = {
      ...userEnvironment,
      OPENCODE_CONFIG_DIR: this.opencodeConfigDir,
      OPENCODE_DB: dbPath,
      EMBEDDEDCOWORK_INSTANCE_ID: id,
      EMBEDDEDCOWORK_BASE_URL: this.options.getServerBaseUrl(),
      ...(this.options.nodeExtraCaCertsPath ? { NODE_EXTRA_CA_CERTS: this.options.nodeExtraCaCertsPath } : {}),
      [OPENCODE_SERVER_USERNAME_ENV]: opencodeAuth.username,
      [OPENCODE_SERVER_PASSWORD_ENV]: opencodeAuth.password,
    }

    const logLevel = (serverConfig as any)?.logLevel

    try {
      const { pid, port, exitPromise, getLastOutput } = await this.runtime.launch({
        workspaceId: id,
        folder: workspacePath,
        binaryPath: resolvedBinaryPath,
        environment,
        logLevel,
        onExit: (info) => this.handleProcessExit(info.workspaceId, info),
      })

      const runtimeVersion = await this.waitForWorkspaceReadiness({ workspaceId: id, port, exitPromise, getLastOutput })
      if (runtimeVersion) {
        descriptor.binaryVersion = runtimeVersion
      }

      descriptor.pid = pid
      descriptor.port = port
      descriptor.status = "ready"
      descriptor.updatedAt = new Date().toISOString()
      this.options.eventBus.publish({ type: "workspace.started", workspace: descriptor })
      this.options.logger.info({ workspaceId: id, port }, "Workspace ready")
      return descriptor
    } catch (error) {
      descriptor.status = "error"
      descriptor.error = error instanceof Error ? error.message : String(error)
      descriptor.updatedAt = new Date().toISOString()
      this.options.eventBus.publish({ type: "workspace.error", workspace: descriptor })
      this.options.logger.error({ workspaceId: id, err: error }, "Workspace failed to start")
      throw error
    }
  }

  async delete(id: string): Promise<WorkspaceDescriptor | undefined> {
    const workspace = this.workspaces.get(id)
    if (!workspace) return undefined

    this.options.logger.info({ workspaceId: id }, "Stopping workspace")
    const wasRunning = Boolean(workspace.pid)
    if (wasRunning) {
      await this.runtime.stop(id).catch((error) => {
        this.options.logger.warn({ workspaceId: id, err: error }, "Failed to stop workspace process cleanly")
      })
    }

    this.workspaces.delete(id)
    this.opencodeAuth.delete(id)
    clearWorkspaceSearchCache(workspace.path)
    if (!wasRunning) {
      this.options.eventBus.publish({ type: "workspace.stopped", workspaceId: id })
    }
    return workspace
  }

  async shutdown() {
    this.options.logger.info("Shutting down all workspaces")

    const stopTasks: Array<Promise<void>> = []

    for (const [id, workspace] of this.workspaces) {
      if (!workspace.pid) {
        this.options.logger.debug({ workspaceId: id }, "Workspace already stopped")
        continue
      }

      this.options.logger.info({ workspaceId: id }, "Stopping workspace during shutdown")
      stopTasks.push(
        this.runtime.stop(id).catch((error) => {
          this.options.logger.error({ workspaceId: id, err: error }, "Failed to stop workspace during shutdown")
        }),
      )
    }

    if (stopTasks.length > 0) {
      await Promise.allSettled(stopTasks)
    }

    this.workspaces.clear()
    this.opencodeAuth.clear()
    this.options.logger.info("All workspaces cleared")
  }

  private requireWorkspace(id: string): WorkspaceRecord {
    const workspace = this.workspaces.get(id)
    if (!workspace) {
      throw new Error("Workspace not found")
    }
    return workspace
  }

  private resolveBinaryPath(identifier: string): string {
    if (!identifier) {
      return identifier
    }

    const looksLikePath = identifier.includes("/") || identifier.includes("\\") || identifier.startsWith(".")
    if (path.isAbsolute(identifier) || looksLikePath) {
      return identifier
    }

    const locator = process.platform === "win32" ? "where" : "which"

    try {
      const result = spawnSync(locator, [identifier], { encoding: "utf8" })
      if (result.status === 0 && result.stdout) {
        const candidates = result.stdout
          .split(/\r?\n/)
          .map((line) => line.trim())
          .filter((line) => line.length > 0)
          .filter((line) => !/^INFO:/i.test(line))

        if (candidates.length > 0) {
          const resolved = this.pickBinaryCandidate(candidates)
          this.options.logger.debug({ identifier, resolved, candidates }, "Resolved binary path from system PATH")
          return resolved
        }
      } else if (result.error) {
        this.options.logger.warn({ identifier, err: result.error }, "Failed to resolve binary path via locator command")
      }
    } catch (error) {
      this.options.logger.warn({ identifier, err: error }, "Failed to resolve binary path from system PATH")
    }

    const shellResolved = resolveBinaryPathFromUserShell(identifier)
    if (shellResolved) {
      this.options.logger.debug({ identifier, resolved: shellResolved }, "Resolved binary path from user shell")
      return shellResolved
    }

    const installedPath = path.join(BIN_DIR, BINARY_NAME)
    if (existsSync(installedPath)) {
      this.options.logger.debug({ identifier, resolved: installedPath }, "Resolved binary path from installed directory")
      return installedPath
    }

    return identifier
  }

  private pickBinaryCandidate(candidates: string[]): string {
    if (process.platform !== "win32") {
      return candidates[0] ?? ""
    }

    const extensionPreference = [".exe", ".cmd", ".bat", ".ps1"]

    for (const ext of extensionPreference) {
      const match = candidates.find((candidate) => candidate.toLowerCase().endsWith(ext))
      if (match) {
        return match
      }
    }

    return candidates[0] ?? ""
  }

  private async waitForWorkspaceReadiness(params: {
    workspaceId: string
    port: number
    exitPromise: Promise<ProcessExitInfo>
    getLastOutput: () => string
  }): Promise<string | undefined> {

    await Promise.race([
      this.waitForPortAvailability(params.port),
      params.exitPromise.then((info) => {
        throw this.buildStartupError(
          params.workspaceId,
          "exited before becoming ready",
          info,
          params.getLastOutput(),
        )
      }),
    ])

    const version = await this.waitForInstanceHealth(params)

    await Promise.race([
      this.delay(STARTUP_STABILITY_DELAY_MS),
      params.exitPromise.then((info) => {
        throw this.buildStartupError(
          params.workspaceId,
          "exited shortly after start",
          info,
          params.getLastOutput(),
        )
      }),
    ])

    return version
  }

  private async waitForInstanceHealth(params: {
    workspaceId: string
    port: number
    exitPromise: Promise<ProcessExitInfo>
    getLastOutput: () => string
  }): Promise<string | undefined> {
    const probeResult = await Promise.race([
      this.probeInstance(params.workspaceId, params.port),
      params.exitPromise.then((info) => {
        throw this.buildStartupError(
          params.workspaceId,
          "exited during health checks",
          info,
          params.getLastOutput(),
        )
      }),
    ])

    if (probeResult.ok) {
      return probeResult.version
    }

    const latestOutput = params.getLastOutput().trim()
    if (latestOutput) {
      throw new Error(latestOutput)
    }
    const reason = probeResult.reason ?? "Health check failed"
    throw new Error(`Workspace ${params.workspaceId} failed health check: ${reason}.`)
  }

  private async probeInstance(
    workspaceId: string,
    port: number,
  ): Promise<{ ok: boolean; reason?: string; version?: string }> {
    const url = `http://127.0.0.1:${port}/global/health`

    try {
      const headers: Record<string, string> = {}
      const authHeader = this.opencodeAuth.get(workspaceId)?.authorization
      if (authHeader) {
        headers["Authorization"] = authHeader
      }

      const response = await fetch(url, { headers })
      if (!response.ok) {
        const reason = `/global/health returned HTTP ${response.status}`
        this.options.logger.debug({ workspaceId, status: response.status }, "Health probe returned server error")
        return { ok: false, reason }
      }

      const payload = (await response.json().catch(() => null)) as null | { healthy?: unknown; version?: unknown }
      const healthy = payload?.healthy === true
      const version = typeof payload?.version === "string" ? payload.version.trim() : undefined

      if (!healthy) {
        const reason = "Instance reported unhealthy"
        this.options.logger.debug({ workspaceId, payload }, "Health probe returned unhealthy response")
        return { ok: false, reason }
      }

      return { ok: true, version: version || undefined }
    } catch (error) {
      const reason = error instanceof Error ? error.message : String(error)
      this.options.logger.debug({ workspaceId, err: error }, "Health probe failed")
      return { ok: false, reason }
    }
  }

  private buildStartupError(
    workspaceId: string,
    phase: string,
    exitInfo: ProcessExitInfo,
    lastOutput: string,
  ): Error {
    const exitDetails = this.describeExit(exitInfo)
    const trimmedOutput = lastOutput.trim()
    const outputDetails = trimmedOutput ? ` Last output: ${trimmedOutput}` : ""
    return new Error(`Workspace ${workspaceId} ${phase} (${exitDetails}).${outputDetails}`)
  }

  private waitForPortAvailability(port: number, timeoutMs = 5000): Promise<void> {
    return new Promise((resolve, reject) => {
      const deadline = Date.now() + timeoutMs
      let settled = false
      let retryTimer: NodeJS.Timeout | null = null

      const cleanup = () => {
        settled = true
        if (retryTimer) {
          clearTimeout(retryTimer)
          retryTimer = null
        }
      }

      const tryConnect = () => {
        if (settled) {
          return
        }
        const socket = connect({ port, host: "127.0.0.1" }, () => {
          cleanup()
          socket.end()
          resolve()
        })
        socket.once("error", () => {
          socket.destroy()
          if (settled) {
            return
          }
          if (Date.now() >= deadline) {
            cleanup()
            reject(new Error(`Workspace port ${port} did not become ready within ${timeoutMs}ms`))
          } else {
            retryTimer = setTimeout(() => {
              retryTimer = null
              tryConnect()
            }, 100)
          }
        })
      }

      tryConnect()
    })
  }

  private delay(durationMs: number): Promise<void> {
    if (durationMs <= 0) {
      return Promise.resolve()
    }
    return new Promise((resolve) => setTimeout(resolve, durationMs))
  }

  private describeExit(info: ProcessExitInfo): string {
    if (info.signal) {
      return `signal ${info.signal}`
    }
    if (info.code !== null) {
      return `code ${info.code}`
    }
    return "unknown reason"
  }

  private handleProcessExit(workspaceId: string, info: { code: number | null; requested: boolean }) {
    const workspace = this.workspaces.get(workspaceId)
    if (!workspace) return

    this.opencodeAuth.delete(workspaceId)

    this.options.logger.info({ workspaceId, ...info }, "Workspace process exited")

    workspace.pid = undefined
    workspace.port = undefined
    workspace.updatedAt = new Date().toISOString()

    if (info.requested || info.code === 0) {
      workspace.status = "stopped"
      workspace.error = undefined
      this.options.eventBus.publish({ type: "workspace.stopped", workspaceId })
    } else {
      workspace.status = "error"
      workspace.error = `Process exited with code ${info.code}`
      this.options.eventBus.publish({ type: "workspace.error", workspace })
    }
  }
}

File: packages/server-rust/src/workspaces/manager.rs
use std::collections::HashMap;
use std::path::Path;
use std::sync::Arc;
use tokio::sync::Mutex;
use tokio::time::{sleep, Duration};

use sha2::{Digest, Sha256};

use crate::api_types::{
    FileSystemEntry, FileSystemEntryType, WorkspaceDescriptor, WorkspaceEventPayload,
    WorkspaceFileResponse, WorkspaceStatus,
};
use crate::events::bus::EventBus;
use crate::logger::Logger;
use crate::opencode_config::OpenCodeConfig;
use crate::settings::binaries::BinaryResolver;
use crate::workspaces::opencode_auth::{
    build_opencode_basic_auth_header, generate_opencode_server_password,
    DEFAULT_OPENCODE_USERNAME, OPENCODE_SERVER_PASSWORD_ENV, OPENCODE_SERVER_USERNAME_ENV,
};
use crate::workspaces::runtime::{LaunchOptions, WorkspaceRuntime};

pub struct WorkspaceManager {
    workspaces: Arc<Mutex<HashMap<String, WorkspaceRecord>>>,
    runtime: WorkspaceRuntime,
    root_dir: String,
    event_bus: EventBus,
    binary_resolver: BinaryResolver,
    #[allow(dead_code)]
    settings: SettingsService,
    #[allow(dead_code)]
    logger: Logger,
    opencode_auth: Arc<Mutex<HashMap<String, OpenCodeAuthEntry>>>,
    /// The server's own base URL (e.g. http://127.0.0.1:18081).
    /// Set after the server starts listening. Uses std::sync::Mutex because
    /// it is set once during startup from a sync context and only read later.
    server_base_url: Arc<std::sync::Mutex<Option<String>>>,
}

use crate::settings::service::SettingsService;

#[derive(Clone)]
struct WorkspaceRecord {
    id: String,
    path: String,
    name: Option<String>,
    status: WorkspaceStatus,
    pid: Option<u32>,
    port: Option<u16>,
    proxy_path: String,
    binary_id: String,
    binary_label: String,
    binary_version: Option<String>,
    created_at: String,
    updated_at: String,
    error: Option<String>,
}

#[derive(Clone)]
struct OpenCodeAuthEntry {
    #[allow(dead_code)]
    username: String,
    #[allow(dead_code)]
    password: String,
    authorization: String,
}

impl WorkspaceManager {
    pub fn new(
        root_dir: String,
        settings: SettingsService,
        binary_resolver: BinaryResolver,
        event_bus: EventBus,
        logger: Logger,
    ) -> Self {
        let runtime = WorkspaceRuntime::new(event_bus.clone(), logger.child("runtime"));

        Self {
            workspaces: Arc::new(Mutex::new(HashMap::new())),
            runtime,
            root_dir,
            event_bus,
            binary_resolver,
            settings,
            logger,
            opencode_auth: Arc::new(Mutex::new(HashMap::new())),
            server_base_url: Arc::new(std::sync::Mutex::new(None)),
        }
    }

    /// Set the server's base URL after the server starts listening.
    /// This is used to set EMBEDDEDCOWORK_BASE_URL for the OpenCode process.
    pub fn set_server_base_url(&self, url: &str) {
        if let Ok(mut stored) = self.server_base_url.lock() {
            *stored = Some(url.to_string());
        }
    }

    pub async fn list(&self) -> Vec<WorkspaceDescriptor> {
        let workspaces = self.workspaces.lock().await;
        workspaces.values().map(|w| w.to_descriptor()).collect()
    }

    pub async fn get(&self, id: &str) -> Option<WorkspaceDescriptor> {
        let workspaces = self.workspaces.lock().await;
        workspaces.get(id).map(|w| w.to_descriptor())
    }

    pub async fn get_instance_port(&self, id: &str) -> Option<u16> {
        let workspaces = self.workspaces.lock().await;
        workspaces.get(id).and_then(|w| w.port)
    }

    pub async fn get_instance_authorization_header(&self, id: &str) -> Option<String> {
        let auth = self.opencode_auth.lock().await;
        auth.get(id).map(|e| e.authorization.clone())
    }

    pub async fn create(
        &mut self,
        folder: &str,
        name: Option<String>,
    ) -> Result<WorkspaceDescriptor, String> {
        let workspace_path = if Path::new(folder).is_absolute() {
            folder.to_string()
        } else {
            Path::new(&self.root_dir)
                .join(folder)
                .to_string_lossy()
                .to_string()
        };

        let id = compute_workspace_id(&workspace_path);
        let binary = self.binary_resolver.resolve_default().await;

        // Check that binary exists
        let binary_path = binary.path.clone();
        if !Path::new(&binary_path).exists() {
            return Err(format!(
                "OpenCode binary not found at: {}. Please configure the correct path in settings.",
                binary_path
            ));
        }

        let proxy_path = format!("/workspaces/{}/worktrees/root/instance", id);

        // ── Step 1: Create the workspace record ──
        let record = WorkspaceRecord {
            id: id.clone(),
            path: workspace_path.clone(),
            name,
            status: WorkspaceStatus::Starting,
            pid: None,
            port: None,
            proxy_path,
            binary_id: binary.path.clone(),
            binary_label: binary.label,
            binary_version: binary.version,
            created_at: chrono::Utc::now().to_rfc3339(),
            updated_at: chrono::Utc::now().to_rfc3339(),
            error: None,
        };

        self.workspaces
            .lock()
            .await
            .insert(id.clone(), record.clone());

        self.event_bus.publish(WorkspaceEventPayload::WorkspaceCreated {
            event_type: "workspace.created".to_string(),
            workspace: record.to_descriptor(),
        });

        // ── Step 2: Set up OpenCode basic auth ──
        let username = DEFAULT_OPENCODE_USERNAME.to_string();
        let password = generate_opencode_server_password();
        let authorization =
            build_opencode_basic_auth_header(&username, &password)
                .ok_or_else(|| "Failed to build OpenCode auth header".to_string())?;

        self.opencode_auth.lock().await.insert(
            id.clone(),
            OpenCodeAuthEntry {
                username: username.clone(),
                password: password.clone(),
                authorization: authorization.clone(),
            },
        );

        // ── Step 3: Build environment variables for OpenCode ──
        let mut env = HashMap::new();

        // OpenCode auth
        env.insert(
            OPENCODE_SERVER_USERNAME_ENV.to_string(),
            username,
        );
        env.insert(
            OPENCODE_SERVER_PASSWORD_ENV.to_string(),
            password,
        );

        // OpenCode config directory
        // NOTE: Must use forward slashes on Windows — OpenCode constructs
        // file:// URLs from this path and backslashes produce invalid URLs
        // like "file://%3F\D:\...".
        let opencode_cfg = OpenCodeConfig::new();
        let config_dir_str = opencode_cfg
            .config_dir
            .to_string_lossy()
            .to_string()
            .replace('\\', "/");
        println!(
            "[WORKSPACE] OPENCODE_CONFIG_DIR={}",
            config_dir_str,
        );
        env.insert(
            "OPENCODE_CONFIG_DIR".to_string(),
            config_dir_str,
        );

        // Session database path (project-local mode)
        let db_path = Path::new(&workspace_path)
            .join(".embeddedcowork")
            .join("session")
            .join("data.db");
        if let Some(parent) = db_path.parent() {
            let _ = std::fs::create_dir_all(parent);
        }
        env.insert("OPENCODE_DB".to_string(), db_path.to_string_lossy().to_string());

        // Instance identity
        env.insert("EMBEDDEDCOWORK_INSTANCE_ID".to_string(), id.clone());

        // Server base URL (for callbacks)
        let base_url = self.server_base_url.lock().ok().and_then(|u| u.clone());
        if let Some(ref url) = base_url {
            env.insert("EMBEDDEDCOWORK_BASE_URL".to_string(), url.clone());
        }

        // ── Step 4: Launch OpenCode process ──
        let result = self
            .runtime
            .launch(LaunchOptions {
                workspace_id: id.clone(),
                folder: workspace_path.clone(),
                binary_path,
                environment: Some(env),
                log_level: None,
            })
            .await;

        // ── Step 5: Handle launch result ──
        match result {
            Ok(launch) => {
                // ── Step 5a: Health probe + stability delay ──
                // Aligns with TS server behavior (waitForPortAvailability +
                // waitForInstanceHealth + STARTUP_STABILITY_DELAY_MS).
                let auth_header = self
                    .opencode_auth
                    .lock()
                    .await
                    .get(&id)
                    .map(|e| e.authorization.clone());

                let healthy = Self::probe_instance(launch.port, &auth_header).await;
                sleep(Duration::from_millis(1500)).await;

                let mut workspaces = self.workspaces.lock().await;
                if let Some(record) = workspaces.get_mut(&id) {
                    record.pid = Some(launch.pid);
                    record.port = Some(launch.port);
                    record.status = if healthy {
                        WorkspaceStatus::Ready
                    } else {
                        WorkspaceStatus::Error
                    };
                    record.updated_at = chrono::Utc::now().to_rfc3339();
                    let descriptor = record.to_descriptor();
                    drop(workspaces);

                    if healthy {
                        self.event_bus.publish(
                            WorkspaceEventPayload::WorkspaceStarted {
                                event_type: "workspace.started".to_string(),
                                workspace: descriptor.clone(),
                            },
                        );

                        tracing::info!(
                            component = %self.logger.component,
                            workspace_id = %id,
                            port = %launch.port,
                            pid = %launch.pid,
                            "Workspace ready"
                        );

                        Ok(descriptor)
                    } else {
                        self.event_bus.publish(
                            WorkspaceEventPayload::WorkspaceError {
                                event_type: "workspace.error".to_string(),
                                workspace: descriptor.clone(),
                            },
                        );

                        Err("Instance health check failed".to_string())
                    }
                } else {
                    Err("Workspace was deleted before it finished starting".to_string())
                }
            }
            Err(error) => {
                tracing::error!(
                    component = %self.logger.component,
                    workspace_id = %id,
                    error = %error,
                    "Failed to start OpenCode"
                );

                // Update workspace status to error
                let mut workspaces = self.workspaces.lock().await;
                if let Some(record) = workspaces.get_mut(&id) {
                    record.status = WorkspaceStatus::Error;
                    record.error = Some(error.clone());
                    record.updated_at = chrono::Utc::now().to_rfc3339();
                    let descriptor = record.to_descriptor();
                    drop(workspaces);

                    self.event_bus.publish(WorkspaceEventPayload::WorkspaceError {
                        event_type: "workspace.error".to_string(),
                        workspace: descriptor,
                    });
                }

                Err(error)
            }
        }
    }

    pub async fn delete(&mut self, id: &str) -> Option<WorkspaceDescriptor> {
        let mut workspaces = self.workspaces.lock().await;
        let workspace = workspaces.remove(id)?;

        // Stop the runtime process
        let _ = self.runtime.stop(id).await;

        self.opencode_auth.lock().await.remove(id);

        self.event_bus.publish(WorkspaceEventPayload::WorkspaceStopped {
            event_type: "workspace.stopped".to_string(),
            workspace_id: id.to_string(),
        });

        Some(workspace.to_descriptor())
    }

    pub async fn shutdown(&self) {
        tracing::info!(component = %self.logger.component, "Shutting down all workspaces");
        self.runtime.shutdown().await;
        self.workspaces.lock().await.clear();
        self.opencode_auth.lock().await.clear();
    }

    pub async fn list_files(
        &self,
        workspace_id: &str,
        relative_path: &str,
    ) -> Result<Vec<FileSystemEntry>, String> {
        let workspace_path = {
            let workspaces = self.workspaces.lock().await;
            workspaces
                .get(workspace_id)
                .map(|w| w.path.clone())
                .ok_or_else(|| "Workspace not found".to_string())?
        };

        let dir_path = Path::new(&workspace_path).join(relative_path);
        let mut entries = tokio::fs::read_dir(&dir_path)
            .await
            .map_err(|e| format!("Failed to read directory: {}", e))?;

        let mut result = Vec::new();
        while let Some(entry) = entries
            .next_entry()
            .await
            .map_err(|e| format!("Failed to read entry: {}", e))?
        {
            let metadata = entry
                .metadata()
                .await
                .map_err(|e| format!("Failed to read metadata: {}", e))?;

            let entry_path = entry.path();
            let name = entry.file_name().to_string_lossy().to_string();
            let is_dir = metadata.is_dir();

            let rel_path = entry_path
                .strip_prefix(Path::new(&workspace_path))
                .unwrap_or(&entry_path)
                .to_string_lossy()
                .to_string();

            let absolute_path = entry_path.to_string_lossy().to_string();

            let modified_at = metadata.modified().ok().and_then(|t| {
                let duration = t.duration_since(std::time::UNIX_EPOCH).ok()?;
                chrono::DateTime::from_timestamp(
                    duration.as_secs() as i64,
                    duration.subsec_nanos(),
                )
                .map(|dt| dt.to_rfc3339())
            });

            result.push(FileSystemEntry {
                name,
                path: rel_path,
                absolute_path: Some(absolute_path),
                entry_type: if is_dir {
                    FileSystemEntryType::Directory
                } else {
                    FileSystemEntryType::File
                },
                size: if metadata.is_file() {
                    Some(metadata.len())
                } else {
                    None
                },
                modified_at,
            });
        }

        Ok(result)
    }

    pub async fn read_file(
        &self,
        workspace_id: &str,
        relative_path: &str,
    ) -> Result<WorkspaceFileResponse, String> {
        let workspace_path = {
            let workspaces = self.workspaces.lock().await;
            workspaces
                .get(workspace_id)
                .map(|w| w.path.clone())
                .ok_or_else(|| "Workspace not found".to_string())?
        };

        let full_path = Path::new(&workspace_path).join(relative_path);
        let contents = tokio::fs::read_to_string(&full_path)
            .await
            .map_err(|e| format!("Failed to read file: {}", e))?;

        Ok(WorkspaceFileResponse {
            workspace_id: workspace_id.to_string(),
            relative_path: relative_path.to_string(),
            contents,
        })
    }

    pub async fn write_file(
        &self,
        workspace_id: &str,
        relative_path: &str,
        contents: &str,
    ) -> Result<(), String> {
        let workspace_path = {
            let workspaces = self.workspaces.lock().await;
            workspaces
                .get(workspace_id)
                .map(|w| w.path.clone())
                .ok_or_else(|| "Workspace not found".to_string())?
        };

        let full_path = Path::new(&workspace_path).join(relative_path);
        if let Some(parent) = full_path.parent() {
            tokio::fs::create_dir_all(parent)
                .await
                .map_err(|e| format!("Failed to create parent directories: {}", e))?;
        }

        tokio::fs::write(&full_path, contents)
            .await
            .map_err(|e| format!("Failed to write file: {}", e))?;

        Ok(())
    }

    /// Probe OpenCode instance health endpoint, matching TS server probeInstance().
    /// Returns `true` if the instance responds with `{ healthy: true }`.
    async fn probe_instance(port: u16, auth_header: &Option<String>) -> bool {
        let url = format!("http://127.0.0.1:{}/global/health", port);

        let client = reqwest::Client::builder()
            .timeout(std::time::Duration::from_secs(5))
            .build()
            .ok();

        let client = match client {
            Some(c) => c,
            None => return false,
        };

        let mut req = client.get(&url);
        if let Some(auth) = auth_header {
            req = req.header("Authorization", auth);
        }

        let resp = match req.send().await {
            Ok(r) => r,
            Err(e) => {
                tracing::warn!(%port, error = %e, "Health probe request failed");
                return false;
            }
        };

        if !resp.status().is_success() {
            tracing::warn!(%port, status = %resp.status().as_u16(), "Health probe returned non-success");
            return false;
        }

        let payload: serde_json::Value = match resp.json().await {
            Ok(v) => v,
            Err(e) => {
                tracing::warn!(%port, error = %e, "Health probe response was not valid JSON");
                return false;
            }
        };

        let healthy = payload
            .get("healthy")
            .and_then(|v| v.as_bool())
            .unwrap_or(false);

        if !healthy {
            tracing::warn!(%port, payload = %payload, "Health probe returned unhealthy");
        }

        healthy
    }
}

fn compute_workspace_id(folder: &str) -> String {
    let abs_path =
        std::path::absolute(folder).unwrap_or_else(|_| Path::new(folder).to_path_buf());
    let hash = Sha256::digest(abs_path.to_string_lossy().as_bytes());
    hex::encode(&hash[..6])
}

impl WorkspaceRecord {
    fn to_descriptor(&self) -> WorkspaceDescriptor {
        WorkspaceDescriptor {
            id: self.id.clone(),
            path: self.path.clone(),
            name: self.name.clone(),
            status: self.status.clone(),
            pid: self.pid,
            port: self.port,
            proxy_path: self.proxy_path.clone(),
            binary_id: self.binary_id.clone(),
            binary_label: self.binary_label.clone(),
            binary_version: self.binary_version.clone(),
            created_at: self.created_at.clone(),
            updated_at: self.updated_at.clone(),
            error: self.error.clone(),
        }
    }
}
